### PR TITLE
Custom themes storage

### PR DESCRIPTION
**Pull request by Miodec**
Storing custom themes, based on #2589 with some modifications (because i cant push to his pr for some reason)

### CODE REVIEW

File: backend/api/controllers/user.ts [hunk 1]
```diff
@@ -206,6 +206,45 @@ class UserController {
...
+    const { uid } = req.ctx.decodedToken;
+    const customThemes = await UsersDAO.getThemes(uid);
+    return new MonkeyResponse("Custom themes retrieved", customThemes);
+  }
+
+  static async addCustomTheme(
+    req: MonkeyTypes.Request
+  ): Promise<MonkeyResponse> {
+    const { uid } = req.ctx.decodedToken;
+    const customTheme = req.body;
```
Comment (Bruception) [inline backend/api/controllers/user.ts:221]: While this is okay, I prefer explicitly picking out properties from the request’s body.

File: backend/api/controllers/user.ts [hunk 2]
```diff
@@ -206,6 +206,45 @@ class UserController {
...
+    return new MonkeyResponse("Custom theme added", {
+      theme: addedTheme,
+    });
+  }
+
+  static async removeCustomTheme(
+    req: MonkeyTypes.Request
+  ): Promise<MonkeyResponse> {
+    const { uid } = req.ctx.decodedToken;
+    const { themeId } = req.body;
```
Comment (Bruception) [inline backend/api/controllers/user.ts:233]: ^

File: backend/api/routes/users.ts [hunk 3]
```diff
@@ -20,6 +20,44 @@ const tagNameValidation = joi
...
+      "Name cannot contain special characters. Can include _ . and -",
+  });
+
+const customThemeColorsValidation = joi
+  .array()
+  .items(
+    joi
+      .string()
+      .length(7)
+      .regex(/^[#][0-9a-fA-F]+$/)
```
Comment (Bruception) [inline backend/api/routes/users.ts:40]: This regex should be simplified/fixed.

File: backend/api/routes/users.ts [hunk 4]
```diff
@@ -20,6 +20,44 @@ const tagNameValidation = joi
...
+
+const customThemeIdValidation = joi
+  .string()
+  .length(24)
+  .regex(/^[0-9a-fA-F]+$/)
+  .required()
+  .messages({
+    "string.length": "The themeId must be 24 characters long",
+    "string.pattern.base": "The themeId must be valid hexadecimal string",
+  });
```
Comment (Bruception) [inline backend/api/routes/users.ts:60]: New line here

File: backend/dao/user.js [hunk 5]
```diff
@@ -83,7 +83,7 @@ class UsersDAO {
   }
 
   static async addTag(uid, name) {
-    let _id = new ObjectId();
+    const _id = new ObjectId();
```
Comment (Bruception) [inline backend/dao/user.js:86]: I don’t think this is needed. Creation of tag document will return an id.

File: backend/dao/user.js [hunk 6]
```diff
@@ -298,6 +298,89 @@ class UsersDAO {
...
+      name: theme.name,
+    };
+  }
+
+  static async removeTheme(uid, _id) {
+    const user = await db.collection("users").findOne({ uid });
+    if (!user)
+      throw new MonkeyError(404, "User not found", "Remove custom theme");
+
+    if (this.themeExists(user.customThemes, _id))
```
Comment (Bruception) [inline backend/dao/user.js:339]: This is confusing

File: backend/dao/user.js [hunk 7]
```diff
@@ -298,6 +298,89 @@ class UsersDAO {
...
+  }
+
+  static async getThemes(uid) {
+    const user = await db.collection("users").findOne({ uid });
+    if (!user)
+      throw new MonkeyError(404, "User not found", "Get custom themes");
+    return user.customThemes ?? [];
+  }
+
+  static async setApeKeys(uid, apeKeys) {
```
Comment (Bruception) [inline backend/dao/user.js:380]: This needs to be deleted. We should rebase this properly

File: frontend/src/scripts/config.ts [hunk 8]
```diff
@@ -1612,13 +1629,16 @@ export function apply(
     }
   );
   if (configObj !== undefined && configObj !== null) {
-    setCustomThemeColors(configObj.customThemeColors, true);
+    // setCustomThemeColors(configObj.customThemeColors, true);
```
Comment (Bruception) [inline frontend/src/scripts/config.ts:1632]: Remove unused commented code

File: frontend/src/scripts/config.ts [hunk 9]
```diff
@@ -1612,13 +1629,16 @@ export function apply(
...
     setThemeLight(configObj.themeLight, true);
     setThemeDark(configObj.themeDark, true);
     setAutoSwitchTheme(configObj.autoSwitchTheme, true);
-    setThemes(configObj.theme, configObj.customTheme, true);
-    // setTheme(configObj.theme, true);
```
Comment (Bruception) [inline frontend/src/scripts/config.ts:1620]: Same here

File: frontend/src/scripts/types/types.d.ts [hunk 10]
```diff
@@ -258,7 +267,7 @@ declare namespace MonkeyTypes {
     themeLight: string;
     themeDark: string;
     autoSwitchTheme: boolean;
-    customTheme: boolean;
+    customTheme: bool;
```
Comment (Bruception) [inline frontend/src/scripts/types/types.d.ts:270]: Bool?

File: backend/dao/user.js (hunk 5 above)
Comment (Miodec) [reply backend/dao/user.js:86]: Unrelated to the pr. Will fix outside

File: backend/api/routes/users.ts [hunk 11]
```diff
@@ -20,6 +20,45 @@ const tagNameValidation = joi
...
+      "Name cannot contain special characters. Can include _ . and -",
+  });
+
+const customThemeColorsValidation = joi
+  .array()
+  .items(
+    joi
+      .string()
+      .length(7)
+      .regex(/^#(?:[0-9a-fA-F]{6})$/)
```
Comment (Bruception) [inline backend/api/routes/users.ts:40]: This non capturing group operator is not doing anything, is fine `^#[0-9a-fA-F]{6}$`

**Pull request merged**